src/metabase/models/action.clj

```diff
@@ -203,41 +211,43 @@
-                                  (let [implicit-params (get implicit-parameters-by-model-id model-id)
-                                        saved-params    (m/index-by :id parameters)]
-                                    (for [param implicit-params
-                                          :let [saved-param (get saved-params (:id param))]]
-                                      (merge param saved-param))))]]
-      (cond-> action
-        implicit-params
-        (m/assoc-some :parameters (cond->> implicit-params
-                                    (= "row/delete" (:kind action))
-                                    (filter ::pk?)
-
-                                    (contains? #{"row/update" "row/delete"} (:kind action))
-                                    (map (fn [param] (cond-> param (::pk? param) (assoc :required true))))
-
-                                    :always
-                                    (map #(dissoc % ::pk? ::field-id))
-
-                                    :always seq))))))
+  (let [actions                       (apply select-actions-without-implicit-params options)
+        implicit-action-model-ids     (set (map :model_id (filter (comp #(= :implicit %) :type) actions)))
+        implicit-action-models        (if known-models
+                                        (->> known-models
+                                             (filter #(contains? implicit-action-model-ids (:id %)))
+                                             distinct)
+                                        (when (seq implicit-action-model-ids)
+                                          (db/select 'Card :id [:in implicit-action-model-ids])))
+        model-id->db-id               (into {} (for [card implicit-action-models]
+                                                 [(:id card) (:database_id card)]))
+        model-id->implicit-parameters (when (seq implicit-action-models)
+                                        (implicit-action-parameters implicit-action-models))]
+    (for [{:keys [parameters] :as action} actions]
+      (if (= (:type action) :implicit)
+        (let [model-id        (:model_id action)
+              saved-params    (m/index-by :id parameters)
+              implicit-params (cond->> (get model-id->implicit-parameters model-id)
+                                :always
+                                (map (fn [param] (merge param (get saved-params (:id param)))))
+
+                                (= "row/delete" (:kind action))
+                                (filter ::pk?)
+
+                                (contains? #{"row/update" "row/delete"} (:kind action))
+                                (map (fn [param] (cond-> param (::pk? param) (assoc :required true))))
+
+                                :always
+                                (map #(dissoc % ::pk? ::field-id)))]
+          (cond-> (assoc action :database_id (model-id->db-id (:model_id action)))
```

Comment:
This function was screaming out for a refactor, so I tidied it up a little. 

But this line, and lines 227-228 are the only real changes. In english, this adds the `:database_id` from the model if it's an implicit action.

Reply:
Seems like bits of it, especially L233-246, would be a lot clearer if they were pulled out into sensibly-named functions. This is a 32-line function, and it'd be great to have something like four 8-line functions instead.

Reply:
Refactoring PR coming up...